pkg/kv/kvserver/uncertainty/doc.go

```diff
@@ -192,6 +194,116 @@ var D6 = roachpb.Transaction{}.ObservedTimestamps
+//    local HLC clock to ensure that it considers any value that was part of a
+//    transaction which could have committed before the request was received by
+//    the leaseholder to be uncertain. Concretely, the non-transactional request
+//    needs to consider values of the following form to be uncertain:
+//
+//      written_timestamp < local_limit && commit_timestamp < global_limit
+//
+//    The value that the non-transactional request is observing may have been
+//    written on the local leaseholder at time 10, its transaction may have been
+//    committed remotely at time 20, acknowledged, then the non-transactional
+//    request may have begun and received a timestamp of 15 from the local
+//    leaseholder, then finally the value may have been resolved asynchronously
+//    and moved to timestamp 20 (written_timestamp: 10, commit_timestamp: 20).
+//    The failure of the non-transactional request to observe this value would
+//    be a stale read.
+//
+//    For example, consider the following series of events:
+//    - client A begins a txn and is assigned provisional commit timestamp = 95
+//    - client A's txn performs a Put(k, v1)
+//    - leaseholder serves Put(k, v1), lays down intent at written_timestamp = 95
+//    - client A's txn performs a write elsewhere and hits a WriteTooOldError
+//      that bumps its provisional commit timestamp to 100
+//    - client A's txn refreshes to ts = 100. This notably happens without
+//      involvement of the leaseholder that served the Put (this is at the heart
+//      of #36431), so that leaseholder's clock is not updated
+//    - client A's txn commits remotely and client A receives the acknowledgment
+//    - client B initiates non-txn read of k
+//    - leaseholder assigns read timestamp ts = 97
+//    - asynchronous intent resolution resolves the txn's intent at k, moving v1
+//      to ts = 100 in the process
+//    - non-txn request must use an uncertainty interval that extends past 100
+//      to ensure that it observes k = v1 in uncertainty interval, performs a
+//      server-side retry, bumps its read timestamp, and returns k = v1. Failure
+//      to do so would be a stale read
+//
+//    TODO(nvanbenschoten): expand on this when we fix #36431. For now, this can
+//    be framed in relation to synthetic timestamps, but it's easier to discuss
+//    in terms of the impending "written_timestamp" attribute of each value,
+//    even though written_timestamps do not yet exist in code.
+//
+//    TODO(nvanbenschoten): add more direct testing for this when we fix #36431.
+//
+//    TODO(nvanbenschoten): add another reason here once we address #73292.
+//
+// Convenient, because non-transactional requests are always scoped to a
+// single-range, those that hit uncertainty errors can always retry on the
+// server, so these errors never bubble up to the client that initiated the
+// request.
+var D7 = roachpb.Header{}.TimestampFromServerClock
```

Comment:
Didn't read this comment in detail because I assume it mirrors exactly what is in the PR message and will be synced before merge.